From 34f031e4cb236b58be089bc8ddfe773b41a83fe1 Mon Sep 17 00:00:00 2001
From: "Johnson.Li" <johnson.li@intel.com>
Date: Mon, 22 Aug 2016 15:36:45 +0000
Subject: [PATCH 06/11] Add APIs to set NSH keys for match fields

Signed-off-by: Johnson.Li <johnson.li@intel.com>

diff --git a/include/openvswitch/match.h b/include/openvswitch/match.h
index 60da412..1d8e0ca 100644
--- a/include/openvswitch/match.h
+++ b/include/openvswitch/match.h
@@ -40,6 +40,18 @@ struct match {
 /* Initializer for a "struct match" that matches every packet. */
 #define MATCH_CATCHALL_INITIALIZER { .flow = { .dl_type = 0 } }
 
+#define MATCH_SET_FIELD_MASKED(match, field, value, msk)      \
+    do {                                                      \
+        (match)->wc.masks.field = (msk);                      \
+        (match)->flow.field = (value) & (msk);                \
+    } while (0)
+
+#define MATCH_SET_FIELD_UINT8(match, field, value)            \
+    MATCH_SET_FIELD_MASKED(match, field, value, UINT8_MAX)
+
+#define MATCH_SET_FIELD_BE32(match, field, value)             \
+    MATCH_SET_FIELD_MASKED(match, field, value, OVS_BE32_MAX)
+
 void match_init(struct match *,
                 const struct flow *, const struct flow_wildcards *);
 void match_wc_init(struct match *match, const struct flow *flow);
-- 
1.8.4.2

